zerver/views/message_fetch.py

```diff
@@ -169,74 +170,93 @@ def get_messages_backend(
         assert log_data is not None
         log_data["extra"] = "[{}]".format(",".join(verbose_operators))
 
-    query_info = fetch_messages(
-        narrow=narrow,
-        user_profile=user_profile,
-        realm=realm,
-        is_web_public_query=is_web_public_query,
-        anchor=anchor,
-        include_anchor=include_anchor,
-        num_before=num_before,
-        num_after=num_after,
-    )
-
-    anchor = query_info.anchor
-    include_history = query_info.include_history
-    is_search = query_info.is_search
-    rows = query_info.rows
-
-    # The following is a little messy, but ensures that the code paths
-    # are similar regardless of the value of include_history.  The
-    # 'user_messages' dictionary maps each message to the user's
-    # UserMessage object for that message, which we will attach to the
-    # rendered message dict before returning it.  We attempt to
-    # bulk-fetch rendered message dicts from remote cache using the
-    # 'messages' list.
-    message_ids: List[int] = []
-    user_message_flags: Dict[int, List[str]] = {}
-    if is_web_public_query:
-        # For spectators, we treat all historical messages as read.
-        for row in rows:
-            message_id = row[0]
-            message_ids.append(message_id)
-            user_message_flags[message_id] = ["read"]
-    elif include_history:
-        assert user_profile is not None
-        message_ids = [row[0] for row in rows]
-
-        # TODO: This could be done with an outer join instead of two queries
-        um_rows = UserMessage.objects.filter(user_profile=user_profile, message_id__in=message_ids)
-        user_message_flags = {um.message_id: um.flags_list() for um in um_rows}
+    with transaction.atomic():
+        # We're about to perform a search, and then get results from
+        # it; this is done across multiple queries.  To prevent race
+        # conditions, we want the messages returned to be consistent
+        # with the version of the messages that was searched, to
+        # prevent changes which happened between them from leaking to
+        # clients who should not be able to see the new values, and
+        # when messages are deleted in between.  We set up
+        # repeatable-read isolation for this transaction, so that we
+        # prevent both phantom reads and non-repeatable reads.
+        #
+        # In a read-only repeatable-read transaction, it is not
+        # possible to encounter deadlocks or need retries due to
+        # serialization errors.
+        cursor = connection.cursor()
+        cursor.exectute("SET TRANSACTION ISOLATION LEVEL REPEATABLE READ READ ONLY")
```

Comment:
Is there a way to quote the "repeatable read" and "read only" bits to be more obvious how it'll be parsed?

Reply:
[Unfortunately, no.](https://www.postgresql.org/docs/current/sql-set-transaction.html)